Source/Core/Core/PowerPC/Jit64/Jit.cpp

```diff
@@ -719,9 +720,11 @@ const u8* Jit64::DoJit(u32 em_address, PPCAnalyst::CodeBuffer *code_buf, JitBloc
 		bool gatherPipeIntCheck = js.fifoWriteAddresses.find(ops[i].address) != js.fifoWriteAddresses.end();
 
 		// Gather pipe writes using an immediate address are explicitly tracked.
-		if (jo.optimizeGatherPipe && js.fifoBytesThisBlock >= 32)
+		if (jo.optimizeGatherPipe && (js.fifoBytesThisBlock >= 32 || js.mustCheckFifo))
 		{
-			js.fifoBytesThisBlock -= 32;
+			if (js.fifoBytesThisBlock >= 32)
+				js.fifoBytesThisBlock -= 32;
```

Comment:
I've been a bit deliberate about not changing the existing behaviour in this commit, to be on the safe side. At the moment it should just introduce FIFO checks and not risk moving existing ones.

Moving forward, I'm pretty sure we should never subtract and always do "js.fifoBytesThisBlock = 0". (Supposing a block has written 31 bytes, then writes a 2 byte value, this would generate a check and leave js.fifoBytesThisBlock as 1. Then at block cleanup, it would generate another check, even if no more bytes have been written. I'd prefer to avoid this by setting it to zero and maybe renaming it to something like fifoBytesSinceCheck or pendingFifoBytes.)

I'm very happy to take your guidance though: any preference between the current version, your suggestion, and the always zeroing version?